### DISCUSSION TITLE
note: perhaps two different versions of crate `num_bigint` are being used?

### DISCUSSION THREAD

**Discussion by joao-pedro-braz** [diesel-rs/diesel#3424] [Q&A] [answered]
### Versions

- **Rust:** 1.64
- **Diesel:** 2.0.2
- **Database:** PostgreSQL
- **Operating System** x86_64-unknown-linux-gnu (OpenSUSE Tumbleweed 20221119)

### Feature Flags

- **diesel:** "chrono", "postgres", "numeric", "r2d2", "uuid",  "without-deprecated"

## Problem Description
Error while building when upgrading from Diesel 1.4.8 to 2.0.2:

> error[E0308]: mismatched types
   --> /home/x/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-2.0.2/src/pg/types/numeric.rs:69:42
    |
69  |             let result = BigDecimal::new(BigInt::from_biguint(sign, result), -correction_exp)
    |                          --------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `num_bigint::bigint::BigInt`, found struct `num_bigint::BigInt`
    |                          |
    |                          arguments to this function are incorrect
    |
    = note: perhaps two different versions of crate `num_bigint` are being used?
note: associated function defined here
   --> /home/x/.cargo/registry/src/github.com-1ecc6299db9ec823/bigdecimal-0.1.2/src/lib.rs:149:12
    |
149 |     pub fn new(digits: BigInt, scale: i64) -> BigDecimal {
    |            ^^^

### What are you trying to accomplish?
Trying to update my diesel version from 1.4.8 to 2.0.2

### What is the expected output?
Successful build

### What is the actual output?
Failed build

### Are you seeing any additional errors?
Yes, all related to the one I gave above:

> error[E0308]: mismatched types
   --> /home/x/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-2.0.2/src/pg/types/numeric.rs:113:47
    |
113 |             let mut digits = ToBase10000(Some(integer)).collect::<Vec<_>>();
    |                                          ---- ^^^^^^^ expected struct `BigUint`, found struct `num_bigint::biguint::BigUint`
    |                                          |
    |                                          arguments to this enum variant are incorrect
    |
    = note: perhaps two different versions of crate `num_bigint` are being used?
note: tuple variant defined here
   --> /home/x/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:526:5
    |
526 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^

> error[E0308]: mismatched types
   --> /home/x/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-2.0.2/src/pg/types/numeric.rs:124:17
    |
123 |             match decimal.sign() {
    |                   -------------- this expression has type `num_bigint::bigint::Sign`
124 |                 Sign::Plus => PgNumeric::Positive {
    |                 ^^^^^^^^^^ expected enum `num_bigint::bigint::Sign`, found enum `num_bigint::Sign`
    |
    = note: perhaps two different versions of crate `num_bigint` are being used?

> error[E0308]: mismatched types
   --> /home/x/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-2.0.2/src/pg/types/numeric.rs:129:17
    |
123 |             match decimal.sign() {
    |                   -------------- this expression has type `num_bigint::bigint::Sign`
...
129 |                 Sign::Minus => PgNumeric::Negative {
    |                 ^^^^^^^^^^^ expected enum `num_bigint::bigint::Sign`, found enum `num_bigint::Sign`
    |
    = note: perhaps two different versions of crate `num_bigint` are being used?

> error[E0308]: mismatched types
   --> /home/x/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-2.0.2/src/pg/types/numeric.rs:134:17
    |
123 |             match decimal.sign() {
    |                   -------------- this expression has type `num_bigint::bigint::Sign`
...
134 |                 Sign::NoSign => PgNumeric::Positive {
    |                 ^^^^^^^^^^^^ expected enum `num_bigint::bigint::Sign`, found enum `num_bigint::Sign`
    |
    = note: perhaps two different versions of crate `num_bigint` are being used?

### Steps to reproduce

Might be relevant:

```
[package]
name = "hex-battles-rest-api"
version = "0.1.0"
edition = "2021"

[profile.dev]
opt-level = 1

[profile.dev.package."*"]
opt-level = 3

[dependencies]
serde = { version = "1.0", features = ["derive"] }
gettext = "0.4.0"
gettext-macros = "0.6.1"
gettext-utils = "0.1.0"
derive_more = "0.99.17"
log = "0.4.17"
mailgun_v3 = "0.13.0"
askama = "0.11.1"
jsonwebtoken = "8.1.1"
lazy_static = "1.4.0"
argon2 = "0.4.0"
rand = "0.8.5"
base64 = "0.13.0"
r2d2_redis = "0.14.0"
serde_json = "1.0.81"
rocket_cors = { git = "https://github.com/lawliet89/rocket_cors", branch = "master" }
hex = "0.4.3"
map-macro = "0.2.2"
bat = "0.21.0"

[dependencies.rocket]
version = "0.5.0-rc.2"
features = ["json"]

[dependencies.validator]
version = "0.16.0"
features = ["derive"]

[dependencies.http-api-problem]
version = "0.53.0"
features = ["rocket"]

[dependencies.rocket_i18n]
git = "https://github.com/joao-pedro-braz/rocket_i18n/"
branch = "go-async"
version = "0.5.0"
default-features = false
features = ["rocket"]

[dependencies.chrono]
version = "0.4.19"
features = ["serde"]

[dependencies.diesel]
version = "2.0.2"
features = ["chrono", "postgres", "numeric", "r2d2", "uuid",  "without-deprecated"]

[dependencies.bigdecimal]
version = "0.3.0"
features = ["serde"]

[dependencies.uuid]
version = "0.8.2"
features = ["serde", "v4"]

[dependencies.glam]
version = "0.22.0"
features = ["serde"]

```

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment, marked as answer]
I converted this to a discussion, as our issue tracker is not a support forum. 

Otherwise, the error message likely already answers whats wrong here:

> = note: perhaps two different versions of crate num_bigint are being used?

As that's not reproducible without a minimal example I cannot help anymore than pointing to this part of the error message. (We are running tests on the CI with the `numeric` feature enabled, to be clear here)

**Reply by joao-pedro-braz** [reply]
No problems!
As you pointed out, the problem is probably related to some other dependencies I'm pulling in, gonna have a look there.

Thanks

**Reply by joao-pedro-braz** [reply]
Just getting back to say that I solved it by updating both bigdecimal (to 0.3.0) and uuid (to 1.2.2).
I'm pretty sure I tried that before, but it didn't work. Maybe rustc had something cached and failed to use the newer versions.

Anyway, thanks for the help!

**Reply by weiznich** [reply]
:+1: Thanks for coming back to report that this fixed the problem. This is really helpful for others locking at this discussion to find a solution for similar problems.
